blocks/graph.py

```diff
@@ -32,45 +59,58 @@ def __init__(self, outputs):
-                        recursion(av)
-                    self.updates.extend(application_call.updates)
-            if current.owner:
-                owner = current.owner
-                if owner not in self.applies:
-                    if hasattr(owner.tag, 'updates'):
-                        logger.debug("found updates in application of {}"
-                                     .format(owner))
-                        self.updates.extend(owner.tag.updates.items())
-                    self.applies.add(owner)
-                for input_ in owner.inputs:
-                    if input_ not in self.variables:
-                        recursion(input_)
-
-        for output in self.outputs:
-            if output not in self.variables:
-                recursion(output)
-        self.inputs = [v for v in self.variables if is_graph_input(v)]
-
-    def get_shared_variables(self):
-        """Returns all shared variables found in the computation graph."""
+    @property
+    def inputs(self):
+        """Inputs to the graph, excluding constants and shared variables."""
+        return [var for var in self.variables if is_graph_input(var)]
+
+    @property
+    def intermediary_variables(self):
+        return [var for var in self.variables if
+                var not in self.inputs and
+                var not in self.outputs]
+
+    @property
+    def shared_variables(self):
         return [var for var in self.variables if is_shared_variable(var)]
 
+    @property
+    def auxiliary_variables(self):
+        return [var for var in self.variables if hasattr(var.tag, 'roles') and
+                VariableRole.AUXILIARY in var.tag.roles]
+
+    def _get_variables(self):
+        """Collect variables, updates and auxiliary variables."""
+        updates = OrderedDict()
+
+        # Sort apply nodes topologically, get variables and remove duplicates
+        inputs = graph.inputs(self.outputs)
+        sorted_apply_nodes = sort_apply_nodes([inputs], self.outputs,
+                                              [dependence])
+        seen = set()
+        variables = [var for var in list(chain(
+            *[apply_node.inputs for apply_node in sorted_apply_nodes]))
+            if not (var in seen or seen.add(var))] + self.outputs
+
+        # While preserving order add auxiliary variables, and collect updates
+        i = 0
+        seen = {'application_call': set(), 'annotation': set(), 'brick': set()}
+        while i < len(variables):
+            var = variables[i]
+            for tag in ('application_call', 'annotation', 'brick'):
+                annotation = getattr(var.tag, tag, None)
+                if annotation and annotation not in seen[tag]:
+                    seen[tag].add(annotation)
+                    new_avs = [av for av in annotation.auxiliary_variables
+                               if av not in variables]
+                    variables = variables[:i + 1] + new_avs + variables[i + 1:]
```

Comment:
Although speed is not our top priority, let's avoid quadratic complexity where it can be done at low cost.  I think one will almost never need to select both usual and auxiliary variables in one call, so they can simply be appended to the end of `variables` list. Also when you do a ` if av not in variables` check, please make sure that `variables` is a set.